pkg/serializer/func_serializer.go

```diff
@@ -0,0 +1,84 @@
+// Copyright 2017 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package serializer
+
+var (
+	noRetry = func(i int) bool { return false }
+)
+
+type funcProp struct {
+	f        func() error
+	waitFunc func(int) bool
+}
+
+type funcSerializer struct {
+	funcs chan funcProp
+}
+
+// NewFuncSerializer returns a funcSerializer that will be used to execute
+// functions in the same order they are enqueued.
+func NewFuncSerializer(size uint) *funcSerializer {
+	return &funcSerializer{
+		funcs: make(chan funcProp, size),
+	}
+}
+
+// Run starts the funcSerializer internal worker. It will be stopped once
+// `stopCh` is closed or receives a value.
+func (fs *funcSerializer) Run(stopCh <-chan struct{}) {
+	for {
+		select {
+		case <-stopCh:
+			return
+		case f := <-fs.funcs:
+			retries := 0
+			for {
+				select {
+				case <-stopCh:
+					return
+				default:
+				}
+				retries++
+				if err := f.f(); err != nil {
```

Comment:
We want to log these errors, right?

Reply:
I don't think it's a good idea to do it here since this might be used on different places of the code. When calling the `Enqueue` function it should be done as follow:
```
serNodes.Enqueue(func() error {
    err := d.updateK8sNodeV1(oldK8sNode, newK8sNode)
    if err != nil {
         log.Debugf("this is a dedicated k8s function that failed")
    }
    return err
}, serializer.NoRetry)
```

Reply:
OK, that's fine, but then document that the error will only be used to determine failure if non-nil and that it is the responsibility of the function to log the error.

Reply:
Done.